api/v1alpha1/gateway_types.go

```diff
@@ -54,78 +57,185 @@ type GatewayList struct {
+
+	// Address requested for this gateway. This is optional and behavior
+	// can depend on the GatewayClass. If a value is set in the spec and
+	// the requested address is invalid, the GatewayClass MUST indicate
+	// this in the associated entry in GatewayStatus.Listeners. If no
+	// addresses are specified, the GatewayClass may schedule the Gateway
+	// in an implementation-defined manner, providing it with an appropriate
+	// set of addresses.
 	//
-	// Support: Core
+	// Support:
 	//
-	Routes RouteBindingSelector `json:"routes" protobuf:"bytes,3,opt,name=routes"`
+	// +optional
+	Addresses []ListenerAddress `json:"endpoints" protobuf:"bytes,3,rep,name=endpoints"`
 }
 
+// ProtocolType defines the application protocol accepted by a Listener.
+type ProtocolType string
+
 const (
-	// HTTPProtocol constant.
-	HTTPProtocol = "HTTP"
-	// HTTPSProtocol constant.
-	HTTPSProtocol = "HTTPS"
+	// HTTPProtocolType accepts cleartext HTTP/1.1 sessions over TCP.
+	HTTPProtocolType ProtocolType = "HTTP"
+
+	// HTTPSProtocolType accepts HTTP/1.1 or HTTP/2 sessions over TLS.
+	HTTPSProtocolType ProtocolType = "HTTPS"
+
+	// TLSProtocolType accepts TLS sessions over TCP.
+	TLSProtocolType ProtocolType = "TLS"
+
+	// TCPProtocolType accepts TCP sessions.
+	TCPProtocolType ProtocolType = "TCP"
 )
 
-// Listener defines a
+// HostnameMatchType specifies the types of matches that are valid
+// for host names.
+type HostnameMatchType string
+
+const (
+	// HostnameMatchDomain specifies that the host name provided by the
+	// client should be matched against a DNS domain value. The domain
+	// match should follow the wildcard certificate matching rules in
+	// [RFC 6125](https://tools.ietf.org/html/rfc6125#section-6.4.3).
+	HostnameMatchDomain HostnameMatchType = "Domain"
+
+	// HostnameMatchExact specifies that the host name provided by the
+	// client must exactly match the specified value.
+	HostnameMatchExact HostnameMatchType = "Exact"
+)
+
+// HostnameMatch specifies how a Listener should match the host name from a
+// client request. Depending on the incoming protocol, the match may apply to
+// the TLS and the HTTP protocol.
+type HostnameMatch struct {
+	// Match specifies how the host name provided by the client should be
+	// matched against the given value.
+	//
+	// +optional
+	// +kubebuilder:validation:Enum=Domain;Exact
+	// +kubebuilder:default=Exact
+	Match HostnameMatchType `json:"match" protobuf:"bytes,1,name=match"`
+
+	// Value contains the value to match against. This value must be a
+	// fully qualified host or domain name as defined by
+	// [RFC 3986](https://tools.ietf.org/html/rfc3986#section-3.2.2).
+	// Note the following deviations from the "host" part of the URI as defined in the RFC:
+	//
+	// 1. IPs are not allowed.
+	// 2. The `:` delimiter is not respected because ports are not allowed.
+	//
+	// +required
+	// +kubebuilder:validation:MinLength=1
+	Value string `json:"value" protobuf:"bytes,2,name=value"`
```

Comment:
Is `name` a better name here? `value` seems way too generic.

Reply:
Sold!